core/src/main/java/io/kestra/core/runners/Worker.java

```diff
@@ -770,9 +771,17 @@ private boolean waitForTasksCompletion(final Duration timeout) {
 
                     // wait for remaining tasks termination.
                     remaining = Math.max(0, Instant.now().until(deadline, ChronoUnit.MILLIS));
-                    if (remaining > 0) {
-                        this.executorService.awaitTermination(remaining, TimeUnit.MILLISECONDS);
+                    if (remaining == 0) {
+                        return;
+                    }
+
+                    boolean gracefullyShutdown = this.executorService.awaitTermination(remaining, TimeUnit.MILLISECONDS);
+                    if (!gracefullyShutdown) {
+                        log.warn("Worker still has some pending threads after `terminationGracePeriod`. Forcing shutdown now.");
+                        this.executorService.shutdownNow();
                     }
+
+                    shutdownState.set(gracefullyShutdown ? TERMINATED_GRACEFULLY : TERMINATED_FORCED);
                 } catch (InterruptedException e) {
                     log.error("Failed to shutdown the worker. Thread was interrupted");
```

Comment:
we need to add shutdownState.set(TERMINATED_FORCED) in the case the shutdown-thread is interrupted by the JVM.